File: javascript/codeql/path_injection/benign_edge.js
const express = require("express");
const fs = require("fs");
const path = require("path");

const app = express();
const baseDir = path.resolve("/srv/app/uploads");

app.get("/read-contained", (req, res) => {
  const resolved = path.resolve(baseDir, String(req.query.name || "welcome.txt"));
  if (!resolved.startsWith(baseDir + path.sep)) {
    res.status(400).send("blocked");
    return;
  }
  res.send(fs.readFileSync(resolved, "utf8"));
});

File: python/codeql/path_injection/benign_edge.py
from pathlib import Path

from flask import Flask, request


app = Flask(__name__)
BASE_DIR = Path("/srv/app/uploads").resolve()


@app.get("/read-contained")
def read_contained():
    name = request.args.get("name", "welcome.txt")
    path = (BASE_DIR / name).resolve()
    if BASE_DIR not in path.parents:
        return "blocked", 400
    return path.read_text(encoding="utf-8")
